From efcf9bd3671891e8bae87e24c45224936a5ddcf8 Mon Sep 17 00:00:00 2001
From: arathinx <ayyappanx.rathinam@intel.com>
Date: Wed, 4 Nov 2020 17:22:35 +0800
Subject: [PATCH] DRVLIB_SW-3918: I2C driver stuck during continuous bootup

In bootup, lm75 0-004a sensor driver passing the 'i2c_msg->len'
zero to the i2c driver ltq_i2c_xfer() call for TX write four
times consecutively for the addresses 0x4c, 0x4d, 0x4e, 0x4f.
This causes the system stuck in i2c driver.
Added the 'i2c_msg->len >0' check in the i2c driver to avoid this issue"
---
 drivers/i2c/busses/i2c-lantiq.c | 67 +++++++++++++++++++++--------------------
 1 file changed, 35 insertions(+), 32 deletions(-)

diff --git a/drivers/i2c/busses/i2c-lantiq.c b/drivers/i2c/busses/i2c-lantiq.c
index 0664b2123af1..5c9b337845e9 100644
--- a/drivers/i2c/busses/i2c-lantiq.c
+++ b/drivers/i2c/busses/i2c-lantiq.c
@@ -524,42 +524,45 @@ static int ltq_i2c_xfer(struct i2c_adapter *adap, struct i2c_msg msgs[],
 	}
 
 	while (priv->msgs_num) {
-		/* start the transfers */
-		ltq_i2c_xfer_init(priv);
-		/* wait for transfers to complete */
-		ret = wait_for_completion_interruptible_timeout(
-			&priv->cmd_complete, LTQ_I2C_XFER_TIMEOUT);
-		if (ret == 0) {
-			i2c_recover_bus(adap);
-			ltq_i2c_hw_init(adap);
-			dev_err(priv->dev, "controller timed out 0x%x\n",
-				priv->msg_err);
-			ret = -ETIMEDOUT;
-			goto done;
-		} else if (ret < 0)
-			goto done;
-
-		if (priv->msg_err) {
-			if (priv->msg_err & LTQ_I2C_NACK)
-				ret = -ENXIO;
-			else {
-				dev_err(priv->dev, "xfer error:\n");
-				if (priv->msg_err & LTQ_I2C_TX_OFL)
-					dev_err(priv->dev, " tx overflow\n");
-				if (priv->msg_err & LTQ_I2C_TX_UFL)
-					dev_err(priv->dev, " tx underflow\n");
-				if (priv->msg_err & LTQ_I2C_RX_OFL)
-					dev_err(priv->dev, " rx overflow\n");
-				if (priv->msg_err & LTQ_I2C_RX_UFL)
-					dev_err(priv->dev, " rx underflow\n");
-
-				ret = -EREMOTEIO;
+		if (priv->current_msg->len > 0) {
+			/* start the transfers */
+			ltq_i2c_xfer_init(priv);
+			/* wait for transfers to complete */
+			ret = wait_for_completion_interruptible_timeout(
+				&priv->cmd_complete, LTQ_I2C_XFER_TIMEOUT);
+			if (ret == 0) {
 				i2c_recover_bus(adap);
 				ltq_i2c_hw_init(adap);
-			}
+				dev_err(priv->dev, "controller timed out 0x%x\n",
+					priv->msg_err);
+				ret = -ETIMEDOUT;
+				goto done;
+			} else if (ret < 0)
+				goto done;
+
+			if (priv->msg_err) {
+				if (priv->msg_err & LTQ_I2C_NACK)
+					ret = -ENXIO;
+				else {
+					dev_err(priv->dev, "xfer error:\n");
+					if (priv->msg_err & LTQ_I2C_TX_OFL)
+						dev_err(priv->dev, " tx overflow\n");
+					if (priv->msg_err & LTQ_I2C_TX_UFL)
+						dev_err(priv->dev, " tx underflow\n");
+					if (priv->msg_err & LTQ_I2C_RX_OFL)
+						dev_err(priv->dev, " rx overflow\n");
+					if (priv->msg_err & LTQ_I2C_RX_UFL)
+						dev_err(priv->dev, " rx underflow\n");
+
+					ret = -EREMOTEIO;
+					i2c_recover_bus(adap);
+					ltq_i2c_hw_init(adap);
+				}
 
-			goto done;
+				goto done;
+			}
 		}
+
 		if (--priv->msgs_num)
 			priv->current_msg++;
 	}
